Include/internal/pycore_stackref.h

```diff
@@ -0,0 +1,189 @@
+#ifndef Py_INTERNAL_STACKREF_H
+#define Py_INTERNAL_STACKREF_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include <stddef.h>
+#include "pycore_gc.h"
+
+// Mark an object as supporting deferred reference counting. This is a no-op
+// in the default (with GIL) build. Objects that use deferred reference
+// counting should be tracked by the GC so that they are eventually collected.
+extern void _PyObject_SetDeferredRefcount(PyObject *op);
+
+static inline int
+_PyObject_HasDeferredRefcount(PyObject *op)
+{
+#ifdef Py_GIL_DISABLED
+    return (op->ob_gc_bits & _PyGC_BITS_DEFERRED) != 0;
+#else
+    return 0;
+#endif
+}
+
+typedef union {
+    uintptr_t bits;
+} _PyStackRef;
+
+static const _PyStackRef Py_STACKREF_NULL = { .bits = 0 };
+
+#ifdef Py_GIL_DISABLED
+    #define Py_TAG_DEFERRED (1)
+    #define Py_TAG (Py_TAG_DEFERRED)
+#else
+    #define Py_TAG 0
+#endif
```

Comment:
Given the functions below, `Py_TAG` doesn't seem necessary. Can we just define `Py_TAG_DEFERRED` unconditionally? 

It seems reasonable to have asserts like the following in the default build:

```
assert((tagged.bits & Py_TAG_DEFERRED) == 0);
```

Reply:
I can't insert the relevant assert in the default build because we cannot always use static inline functions for hot stuff like this. There have been issues in the recent past where MSVC does not inline the function, so we do actually generate worse code. I don't know if that has been solved, but I would be cautious considering how often this no-op will turn up in the default build.